### PR TITLE
#176 User Management: Adding Users

### PR DESCRIPTION
**Pull request by a2batic**
Bug-id: #176
Signed-off-by: a2batic <murarkakanika@gmail.com>

### CODE REVIEW

**Comment by a2batic** [comment]
@anivargi @brainfunked @cloudbehl @julienlim @gnehapk, please review

File: specs/Spec_for_adding_user.adoc [hunk 1]
```diff
@@ -0,0 +1,166 @@
...
+
+## Problem description
+
+This spec talks about Adding/Creating a new User to Tendrl by the Admin
+
+## Use Cases
+
+The flow goes as follows:-
+
+* A user login as Admin
```
Comment (shtripat) [inline specs/Spec_for_adding_user.adoc]: s/login/logs in/

File: specs/Spec_for_adding_user.adoc [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+### UI changes
+
+* Add *Users* tab in Admin navigation tab
+* Add *Add User* Button in Users Page view
+* Add a form for adding a user. The form will contain following feilds:-
+** User ID
+*** Input type : text
+*** Validations:
+**** Should not contain special characters
```
Comment (shtripat) [inline specs/Spec_for_adding_user.adoc]: special chars like `_`, `-`, '.' should be allowed? Should start with alphanumeric char.

File: specs/Spec_for_adding_user.adoc [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+**** Should be Unique
+**** Can be of a particular Maximum length
```
Comment (shtripat) [inline specs/Spec_for_adding_user.adoc]: There should be a min length check as well

File: specs/Spec_for_adding_user.adoc [hunk 4]
```diff
@@ -0,0 +1,166 @@
...
+
+```
+POST /api/1.0/users - creates a new user
+```
+
+* Sample API response now : 
+
+```
+{
+	"email": "thardy@tendrl.org",
```
Comment (shtripat) [inline specs/Spec_for_adding_user.adoc]: Where is first name and last name here? Is it like both merged in `name` field?

File: specs/Spec_for_adding_user.adoc [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+Need API support for required additional fields:
+
+* first_name
+* last_name
+* email_notifications
+
+
+## Testing:
+
+Test whether fields of adding user form are validating.
```
Comment (shtripat) [inline specs/Spec_for_adding_user.adoc]: Check if user is actually getting created and they are able to login to the system and UI is displayed as per their role :)

File: specs/Spec_for_adding_user.adoc (hunk 1 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: Acknowledge

File: specs/Spec_for_adding_user.adoc (hunk 2 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: Acknowledge, I will elaborate it more

File: specs/Spec_for_adding_user.adoc (hunk 3 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: Yes there is a minimum length too, I will update it

File: specs/Spec_for_adding_user.adoc (hunk 5 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: I will add this :)

File: specs/Spec_for_adding_user.adoc (hunk 4 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: Initially, we just had ```name``` field, now according to new mockup https://redhat.invisionapp.com/share/KNB25OEQT#/screens/226063802, we require separate first name and last name

Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: Or Should we modify the ```name``` only by using JavaScript. What do you suggest @anivargi @gnehapk ?

File: specs/Spec_for_adding_user.adoc (hunk 2 above)
Comment (gnehapk) [reply specs/Spec_for_adding_user.adoc]: @brainfunked @anivargi @nthomas-redhat What validations are required for User ID field? If we want to check for uniqueness then we will require an API call to do that. What should be the minimum and maximum length allowed for this field.

File: specs/Spec_for_adding_user.adoc (hunk 4 above)
Comment (gnehapk) [reply specs/Spec_for_adding_user.adoc]: @a2batic If it is possible for API to provide it, then we can have both fields in UI.  @anivargi Is it doable ?

File: specs/Spec_for_adding_user.adoc [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+* username
+* name
+* role
+
+Required from API - 
+
+* first_name
+* last_name
+* email_notifications
+
```
Comment (gnehapk) [inline specs/Spec_for_adding_user.adoc]: @anivargi Please provide the API response for these fields.

File: specs/Spec_for_adding_user.adoc [hunk 7]
```diff
@@ -0,0 +1,170 @@
...
+* email
+* username
+* name
+* role
+
+Required from API - 
+
+* email_notifications
+
```
Comment (a2batic) [inline specs/Spec_for_adding_user.adoc:112]: @anivargi, please review

File: specs/Spec_for_adding_user.adoc [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+** Confirm Password
+*** Input type : password
+*** Validations:
+**** Should be mandatory
+** Set a Role
+*** Admin
+*** Normal user
+*** Limited user
+*** Input type : Radio button
+
```
Comment (cybernth) [inline specs/Spec_for_adding_user.adoc]: Are we not suporting `Read-only` user?

File: specs/Spec_for_adding_user.adoc [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+*** Validations:
+**** Should be mandatory
+** Last Name
+*** Input type : text
+*** Validations:
+**** Should be mandatory
+** Email
+*** Input type : email
+*** Validations:
+**** Validate email ID
```
Comment (cybernth) [inline specs/Spec_for_adding_user.adoc]: Should this be unique as well?

File: specs/Spec_for_adding_user.adoc [hunk 10]
```diff
@@ -0,0 +1,170 @@
...
+## Proposed change
+
+### UI changes
+
+* Add *Users* tab in Admin navigation tab
+* Add *Add User* Button in Users Page view
+* Add a form for adding a user. The form will contain following feilds:-
+** User ID
+*** Input type : text
+*** Validations:
```
Comment (cybernth) [inline specs/Spec_for_adding_user.adoc:28]: Where all the validations are done? UI or API?

File: specs/Spec_for_adding_user.adoc [hunk 11]
```diff
@@ -0,0 +1,170 @@
...
+**** Should be mandatory
+** Set a Role
+*** Admin
+*** Normal user
+*** Limited user
+*** Input type : Radio button
+
+*NOTE*: 
+
```
Comment (cybernth) [inline specs/Spec_for_adding_user.adoc]: I think we are not going to support enable/disable user. If yes we need to call out explicitly as it is there is in the design

**Review by cybernth — CHANGES_REQUESTED** [review]

File: specs/Spec_for_adding_user.adoc (hunk 10 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc:28]: All validations are done on API only

File: specs/Spec_for_adding_user.adoc (hunk 9 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: @anivargi, Is there any such validation for email field ?

File: specs/Spec_for_adding_user.adoc (hunk 8 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: @nthomas-redhat, According to https://github.com/Tendrl/api/blob/master/docs/authorizaton.adoc , Limited user is same as Read-only user

File: specs/Spec_for_adding_user.adoc (hunk 11 above)
Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: So, should we comment it out from UI for now ?

Comment (a2batic) [reply specs/Spec_for_adding_user.adoc]: @nthomas-redhat, Thanks for reviewing !!

File: specs/Spec_for_adding_user.adoc (hunk 9 above)
Comment (anivargi) [reply specs/Spec_for_adding_user.adoc]: Username and email are unique, checked by the API.

**Pull request merged**